DIPs/DIP1xxx-ajie.md

```diff
@@ -0,0 +1,222 @@
+# Foreach auto ref
+
+| Field           | Value                                                           |
+|-----------------|-----------------------------------------------------------------|
+| DIP:            | (number/id -- assigned by DIP Manager)                          |
+| Review Count:   | 0 (edited by DIP Manager)                                       |
+| Author:         | Ate Eskola (ajieskola@gmail.com)                                |
+| Implementation: | None yet                                                        |
+| Status:         | Draft                                                           |
+
+## Abstract
+
+This DIP proposes that foreach loops should have an option to be annotated
+as `auto ref`, which works for both referrable and non-referrable elements.
```

Comment:
It shouldn't only "work with both [kinds of] elements", it should *infer* their refness. That's different: A `const(int)*` binds `int*` and `immutable(int)*` values, but "forgets" the original mutability, while an `inout(int)` variable infers and keeps mutability information.
In templates, `auto ref` does not only work with both r- and lvalues, it remembers the category. Here it should do exactly the same.
IMO, the current phrasing does not clarify that. DIP 1016 gets you to take-and-forget, so this is beyond DIP 1016.